doc/src/lexer_tutorial/004_external_lib.md

```diff
@@ -172,14 +186,13 @@ impl<'input> Iterator for Lexer<'input> {
   type Item = Spanned<Token, usize, LexicalError>;
 
   fn next(&mut self) -> Option<Self::Item> {
```

Comment:
I'm a bit torn on this `Iterator` impl, It looks like this could just be `Token::lexer(input).spanned().map(...)`, because `lexer()` and `spanned` both return iterators. However I've also run into special cases where a custom `Iterator` implementation was actually necessary.

Reply:
Are you suggesting that I do:

```rust
let lexer = Token::lexer(input).spanned().map(/* ... */);
let parser = ScriptParser::new();
let ast = parser.parse(lexer)?;
```

I tried it locally, and indeed it works well. But I like to encapsulate each part of the process. I could have a function `Token::as_lalrpop(...)` though.

Reply:
Yeah, that is exactly what I was thinking, I also generally do the map through an auxiliary function as well, which I generally named `to_lalrpop_triple`, `as_lalrpop` also seems fine though.  In general do what you think is best/clearest here.

Reply:
In fact, `Token::lexer(input).spanned().map(...)` returns a `std::iter::Map<logos::SpannedIter<'_, Token>, [closure type]>`.

The code becomes quite tricky with the type system, a struct implementing the `Iterator` trait is by far simpler than a function returning the underlying iterator.

Reply:
Using a `struct` allows me to pass data to the implementation of the iterator quite easily compared to function arguments and a separate `to_lalrpop_triple`.

Here I use the filename for error reporting, but you could image an object translating the `Span` into a `(line, col)` tuple. Or anything else.

Thank you for your suggestion, but I think using a struct is simpler. If you're ok with this, I leave things as they are and resolve this conversation.

Reply:
by all means